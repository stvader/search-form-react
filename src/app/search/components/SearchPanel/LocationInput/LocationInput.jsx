import React from 'react';
import { func, array, string } from 'prop-types';
import { TextField, Grid, Typography } from '@material-ui/core';
import Autocomplete from '@material-ui/lab/Autocomplete';
import LocationOnIcon from '@material-ui/icons/LocationOn';
import parse from 'autosuggest-highlight/parse';

const LocationInput = ({ handleChange, handleSelectChange, options, defaultLocation }) => (
  <Autocomplete
    getOptionLabel={option => (typeof option === 'string' ? option : option.description)}
    filterOptions={x => x}
    options={options}
    autoComplete
    includeInputInList
    freeSolo
    defaultValue={defaultLocation}
    disableOpenOnFocus
    onChange={(e, value) => handleSelectChange(e, value)}
    renderInput={params => (
      <TextField
        {...params}
        label="Please add a location"
        variant="outlined"
        fullWidth
        onChange={handleChange}
      />
    )}
    renderOption={option => {
      const matches = option.structured_formatting.main_text_matched_substrings;
      const parts = parse(
        option.structured_formatting.main_text,
        matches.map(match => [match.offset, match.offset + match.length]),
      );

      return (
        <Grid container alignItems="center">
          <Grid item>
            <LocationOnIcon />
          </Grid>
          <Grid item xs>
            {parts.map((part, index) => (
              <span key={index} style={{ fontWeight: part.highlight ? 700 : 400 }}>
                {part.text}
              </span>
            ))}

            <Typography variant="body2" color="textSecondary">
              {option.structured_formatting.secondary_text}
            </Typography>
          </Grid>
        </Grid>
      );
    }}
  />
);

LocationInput.propTypes = {
  handleChange: func,
  handleSelectChange: func,
  options: array,
  defaultLocation: string,
};

export default LocationInput;
