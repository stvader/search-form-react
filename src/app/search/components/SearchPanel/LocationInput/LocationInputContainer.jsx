import React, { memo, useState, useRef, useMemo, useEffect, useContext, useCallback } from 'react';
import throttle from 'lodash/throttle';

import { SearchPanelContext } from '../SearchPanelContext';
import LocationInput from './LocationInput';

import { geocodeByAddress } from '../../../utils/geoCoding';
import { GOOGLE_MAP_LINK } from '../../../search.endpoints';

const loadScript = (src, position, id) => {
  if (!position) {
    return;
  }

  const script = document.createElement('script');
  script.setAttribute('async', '');
  script.setAttribute('id', id);
  script.src = src;
  position.appendChild(script);
};

const autocompleteService = { current: null };

const LocationInputContainer = () => {
  const { selectLocation, handleLocationChange } = useContext(SearchPanelContext);
  const [inputValue, setInputValue] = useState('');
  const [options, setOptions] = useState([]);
  const loaded = useRef(false);

  const defaultLocation = selectLocation && selectLocation.formatted_address;

  if (typeof window !== 'undefined' && !loaded.current) {
    if (!document.querySelector('#google-maps')) {
      loadScript(GOOGLE_MAP_LINK, document.querySelector('head'), 'google-maps');
    }

    loaded.current = true;
  }

  const handleChange = useCallback(
    e => {
      setInputValue(e.target.value);
    },
    [setInputValue],
  );

  const handleSelectChange = useCallback(
    async (e, value) => {
      if (!value) {
        handleLocationChange(null);
        return;
      }

      const address = value && value.description;
      const geocode = await geocodeByAddress({ address });
      const geocodeResult = geocode[0];

      handleLocationChange(geocodeResult);
    },
    [handleLocationChange],
  );

  const fetch = useMemo(
    () =>
      throttle((input, callback) => {
        autocompleteService.current.getPlacePredictions(input, callback);
      }, 200),
    [],
  );

  useEffect(() => {
    let active = true;

    if (!autocompleteService.current && window.google) {
      autocompleteService.current = new window.google.maps.places.AutocompleteService();
    }
    if (!autocompleteService.current) {
      return undefined;
    }

    if (inputValue === '') {
      setOptions([]);
      return undefined;
    }

    fetch({ input: inputValue }, results => {
      if (active) {
        setOptions(results || []);
      }
    });

    return () => {
      active = false;
    };
  }, [inputValue, fetch]);

  return (
    <LocationInput
      options={options}
      handleChange={handleChange}
      handleSelectChange={handleSelectChange}
      defaultLocation={defaultLocation}
    />
  );
};

export default memo(LocationInputContainer);
