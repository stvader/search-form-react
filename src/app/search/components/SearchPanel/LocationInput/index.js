import LocationInput from './LocationInputContainer';

export default LocationInput;
