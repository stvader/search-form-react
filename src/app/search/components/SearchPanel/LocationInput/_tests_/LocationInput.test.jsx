import React from 'react';
import { shallow, configure } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import Autocomplete from '@material-ui/lab/Autocomplete';
import { TextField } from '@material-ui/core';

import LocationInput from '../LocationInput';

configure({ adapter: new Adapter() });

const mockHandleChange = jest.fn();
const mockHandleSelectChange = jest.fn();
const mockOptions = jest.fn();

describe('LocationInput', () => {
  it('LocationInput render', () => {
    const component = shallow(
      <LocationInput
        handleChange={mockHandleChange}
        handleSelectChange={mockHandleSelectChange}
        options={mockOptions}
        currentLocation="Location"
      />,
    );

    const input = component.find(Autocomplete);
    const textField = component.find(TextField);

    expect(component).toMatchSnapshot();
    expect(input).toBeDefined();
    expect(input.prop('defaultValue')).toBe('Location');
    expect(textField).toBeDefined();
  });
});
