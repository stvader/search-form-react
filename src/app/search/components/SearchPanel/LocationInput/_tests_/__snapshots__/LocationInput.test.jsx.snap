// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LocationInput LocationInput render 1`] = `ShallowWrapper {}`;
